sdks/python/apache_beam/io/iobase.py

```diff
@@ -855,24 +864,44 @@ def expand(self, pbegin):
     debug_options = self.pipeline._options.view_as(DebugOptions)
     if debug_options.experiments and 'beam_fn_api' in debug_options.experiments:
       source = self.source
-
-      def split_source(unused_impulse):
-        total_size = source.estimate_size()
-        if total_size:
-          # 1MB = 1 shard, 1GB = 32 shards, 1TB = 1000 shards, 1PB = 32k shards
-          chunk_size = max(1 << 20, 1000 * int(math.sqrt(total_size)))
-        else:
-          chunk_size = 64 << 20  # 64mb
-        return source.split(chunk_size)
-
-      return (
-          pbegin
-          | core.Impulse()
-          | 'Split' >> core.FlatMap(split_source)
-          | util.Reshuffle()
-          | 'ReadSplits' >> core.FlatMap(lambda split: split.source.read(
-              split.source.get_range_tracker(
-                  split.start_position, split.stop_position))))
+      # My original design was to have a SDFBoundedSourceWrapper PTransform, which will
+      # replace Read() when constructing  a pipeline. We can also choose to expand Read
+      # direcly using SDF, which is simpler from my point of view.
+      if 'sdf_bounded_source' in debug_options.experiments:
+        from apache_beam.io.sdf_restriction_provider import SDFBoundedSourceRestrictionProvider
+
+        def create_sdf_bounded_source_dofn():
+          chunk_size = self._get_desired_chunk_size()
```

Comment:
Currently this is being done before pipeline execution but we should move this logic to when the pipeline is executing.